dpkg/scanner.go

```diff
@@ -0,0 +1,208 @@
+	"runtime/trace"
+	"strings"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/internal/scanner"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/tadasv/go-dpkg"
+)
+
+const (
+	name    = "dpkg"
+	kind    = "package"
+	version = "v0.0.1"
+)
+
+var (
+	_ scanner.VersionedScanner = (*Scanner)(nil)
+	_ scanner.PackageScanner   = (*Scanner)(nil)
+)
+
+// Scanner implements the scanner.PackageScanner interface.
+//
+// This looks for directories that look like dpkg databases and examines the
+// "status" file it finds there.
+//
+// The zero value is ready to use.
+type Scanner struct{}
+
+// Name implements scanner.VersionedScanner.
+func (ps *Scanner) Name() string { return name }
+
+// Version implements scanner.VersionedScanner.
+func (ps *Scanner) Version() string { return version }
+
+// Kind implements scanner.VersionedScanner.
+func (ps *Scanner) Kind() string { return kind }
+
+// Scan proxies the call to ScanContext.
+func (ps *Scanner) Scan(layer *claircore.Layer) ([]*claircore.Package, error) {
+	ctx := context.TODO()
+	ctx = log.Logger.WithContext(ctx)
+	return ps.ScanContext(ctx, layer)
+}
+
+// ScanContext attempts to find a dpkg database within the layer and read all of the
+// installed packages it can find in the "status" file.
+//
+// It's expected to return (nil, nil) if there's no dpkg database in the layer.
+//
+// It does not respect any dpkg configuration files.
+func (ps *Scanner) ScanContext(ctx context.Context, layer *claircore.Layer) ([]*claircore.Package, error) {
+	// Preamble
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer:sha256", layer.Hash)
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "package_scanner").
+		Str("name", ps.Name()).
+		Str("version", ps.Version()).
+		Str("kind", ps.Kind()).
+		Str("layer", layer.Hash).
+		Logger()
+	log.Debug().Msg("start")
+	defer log.Debug().Msg("done")
+
+	// Grab a handle to the tarball, make sure we can seek.
+	// If we can't, we'd need another reader for every database found.
+	// It's cleaner to just demand that it's a seeker.
+	rd, err := layer.Reader()
+	if err != nil {
+		log.Warn().Err(err).Msg("opening layer failed")
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+	defer rd.Close()
+	r, ok := rd.(io.ReadSeeker)
+	if !ok {
+		err := errors.New("unable to coerce to io.Seeker")
+		log.Warn().Err(err).Msg("opening layer failed")
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+
+	tr := tar.NewReader(r)
+	// This is a map keyed by directory. A "score" of 2 means this is almost
+	// certainly a dpkg database.
+	loc := make(map[string]int)
+Find:
+	for {
+		h, err := tr.Next()
+		switch err {
+		case nil:
+		case io.EOF:
+			break Find
+		default:
+			log.Warn().Err(err).Msg("reading next header failed")
+			return nil, fmt.Errorf("reading next header failed: %w", err)
+		}
+		switch filepath.Base(h.Name) {
+		case "status", "available":
+			if h.Typeflag == tar.TypeReg {
```

Comment:
edge case: status or available files are symlinks to a different name on the FS. Most likely not worth worrying about but worth mentioning for conversation